File: bind/java/pdfclock.java
/* hello.java
 * Copyright (C) 1997-99 Thomas Merz. All rights reserved.
 *
 * PDFlib client: pdfclock example in Java
 */

import java.text.*;		// SimpleDateFormat
import java.util.*;		// Date

public class pdfclock
{
    public static void main (String argv[])
    {
	long p;
	int tm_hour, tm_min, tm_sec, alpha;
	float RADIUS = 200.0f;
	float MARGIN = 20.0f;
	SimpleDateFormat format;
	Date now = new Date();

	p = pdflib.PDF_new();

	if (pdflib.PDF_open_file(p, "pdfclock_java.pdf") == -1) {
	    System.err.println("Couldn't open PDF file" +"pdfclock_java.pdf\n");
	    System.exit(1);
	}

	pdflib.PDF_set_info(p, "Creator", "pdfclock.java");
	pdflib.PDF_set_info(p, "Author", "Thomas Merz");
	pdflib.PDF_set_info(p, "Title", "PDF clock (Java)");

	pdflib.PDF_begin_page(p,   (int) (2 * (RADIUS + MARGIN)),
				(int) (2 * (RADIUS + MARGIN)));

	pdflib.PDF_set_transition(p, "wipe");
	pdflib.PDF_set_duration(p, 0.5f);

	pdflib.PDF_translate(p, RADIUS + MARGIN, RADIUS + MARGIN);
	pdflib.PDF_setrgbcolor(p, 0.0f, 0.0f, 1.0f);
	pdflib.PDF_save(p);

	// minute strokes 
	pdflib.PDF_setlinewidth(p, 2.0f);
	for (alpha = 0; alpha < 360; alpha += 6)
	{
	    pdflib.PDF_rotate(p, 6.0f);
	    pdflib.PDF_moveto(p, RADIUS, 0.0f);
	    pdflib.PDF_lineto(p, RADIUS-MARGIN/3, 0.0f);
	    pdflib.PDF_stroke(p);
	}

	pdflib.PDF_restore(p);
	pdflib.PDF_save(p);

	// 5 minute strokes
	pdflib.PDF_setlinewidth(p, 3.0f);
	for (alpha = 0; alpha < 360; alpha += 30)
	{
	    pdflib.PDF_rotate(p, 30.0f);
	    pdflib.PDF_moveto(p, RADIUS, 0.0f);
	    pdflib.PDF_lineto(p, RADIUS-MARGIN, 0.0f);
	    pdflib.PDF_stroke(p);
	}

	// how to correctly set the time zone?
	format = new SimpleDateFormat("hh");
	tm_hour= Integer.parseInt(format.format(now));
	format = new SimpleDateFormat("mm");
	tm_min = Integer.parseInt(format.format(now));
	format = new SimpleDateFormat("ss");
	tm_sec = Integer.parseInt(format.format(now));

	// draw hour hand 
	pdflib.PDF_save(p);
	pdflib.PDF_rotate(p, (-((tm_min/60.0f) + tm_hour - 3.0f) * 30.0f));
	pdflib.PDF_moveto(p, -RADIUS/10, -RADIUS/20);
	pdflib.PDF_lineto(p, RADIUS/2, 0.0f);
	pdflib.PDF_lineto(p, -RADIUS/10, RADIUS/20);
	pdflib.PDF_closepath(p);
	pdflib.PDF_fill(p);
	pdflib.PDF_restore(p);

	// draw minute hand
	pdflib.PDF_save(p);
	pdflib.PDF_rotate(p, (-((tm_sec/60.0f) + tm_min - 15.0f) * 6.0f));
	pdflib.PDF_moveto(p, -RADIUS/10, -RADIUS/20);
	pdflib.PDF_lineto(p, RADIUS * 0.8f, 0.0f);
	pdflib.PDF_lineto(p, -RADIUS/10, RADIUS/20);
	pdflib.PDF_closepath(p);
	pdflib.PDF_fill(p);
	pdflib.PDF_restore(p);

	// draw second hand
	pdflib.PDF_setrgbcolor(p, 1.0f, 0.0f, 0.0f);
	pdflib.PDF_setlinewidth(p, 2);
	pdflib.PDF_save(p);
	pdflib.PDF_rotate(p, -((tm_sec - 15.0f) * 6.0f));
	pdflib.PDF_moveto(p, -RADIUS/5, 0.0f);
	pdflib.PDF_lineto(p, RADIUS, 0.0f);
	pdflib.PDF_stroke(p);
	pdflib.PDF_restore(p);

	// draw little circle at center
	pdflib.PDF_circle(p, 0f, 0f, RADIUS/30);
	pdflib.PDF_fill(p);

	pdflib.PDF_restore(p);

	pdflib.PDF_end_page(p);

	pdflib.PDF_close(p);

	pdflib.PDF_delete(p);
    }
}

File: bind/cpp/pdfclock.cpp
/*---------------------------------------------------------------------------*
 |        PDFlib - A library for dynamically generating PDF files            |
 +---------------------------------------------------------------------------+
 |        Copyright (c) 1997-1999 Thomas Merz. All rights reserved.          |
 +---------------------------------------------------------------------------+
 |    This software is not in the public domain.  It is subject to the       |
 |    "Aladdin Free Public License".  See the file license.txt for details.  |
 |    This license grants you the right to use and redistribute PDFlib       |
 |    under certain conditions. Among other things, the license requires     |
 |    that the copyright notice and this notice be preserved on all copies.  |
 |    This requirement extends to ports to other programming languages.      |
 |                                                                           |
 |    In short, you are allowed to develop and use PDFlib-based software     |
 |    as long as you don't sell it. Commercial use of PDFlib requires a      |
 |    commercial license which can be obtained from the author of PDFlib.    |
 |    Contact information can be found in the accompanying PDFlib manual.    |
 |    PDFlib is distributed with no warranty of any kind. Commercial users,  |
 |    however, will receive warranty and support statements in writing.      |
 *---------------------------------------------------------------------------*/

// pdfclock.cpp
//
// A little PDFlib application to draw an analog clock.
//
//

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

#if !defined(WIN32) && !defined(MAC)
#include <unistd.h>
#endif

#include "pdflib.hpp"

#define READMODE	"rb"
#define RADIUS		200.0
#define MARGIN		20.0
#define BUFSIZE		512

#define FILENAME	"pdfclock_cpp.pdf"

int
main()
{
    PDF		*p;
    float	alpha;
    time_t	timer;
    struct tm	ltime;
    
    /* Create a new PDF object */
    p = new PDF();

    /* Open new PDF file */
    if (p->open(FILENAME) == -1) {
	fprintf(stderr, "pdfclock error: cannot open PDF file %s.\n",
		FILENAME);
    }

    p->set_info("Creator", "pdfclock.cpp");
    p->set_info("Author", "Thomas Merz");
    p->set_info("Title", "PDF clock (C++)");

    p->begin_page((unsigned int) (2 * (RADIUS + MARGIN)),
		      (unsigned int) (2 * (RADIUS + MARGIN)));
    
    p->set_transition("wipe");
    p->set_duration(0.5);

    p->translate(RADIUS + MARGIN, RADIUS + MARGIN);
    p->setrgbcolor(0.0, 0.0, 1.0);
    p->save();

    /* minute strokes */
    p->setlinewidth(2.0);
    for (alpha = 0; alpha < 360; alpha += 6)
    {
	p->rotate(6.0);
	p->moveto(RADIUS, 0.0);
	p->lineto((float) (RADIUS-MARGIN/3), 0.0);
	p->stroke();
    }

    p->restore();
    p->save();

    /* 5 minute strokes */
    p->setlinewidth(3.0);
    for (alpha = 0; alpha < 360; alpha += 30)
    {
	p->rotate(30.0);
	p->moveto(RADIUS, 0.0);
	p->lineto(RADIUS-MARGIN, 0.0);
	p->stroke();
    }

    time(&timer);
    ltime = *localtime(&timer);

    /* draw hour hand */
    p->save();
    p->rotate(
	    (float)(-((ltime.tm_min/60.0) + ltime.tm_hour - 3.0) * 30.0));
    p->moveto(-RADIUS/10, -RADIUS/20);
    p->lineto(RADIUS/2, 0.0);
    p->lineto(-RADIUS/10, RADIUS/20);
    p->closepath();
    p->fill();
    p->restore();

    /* draw minute hand */
    p->save();
    p->rotate((float) (-((ltime.tm_sec/60.0) + ltime.tm_min - 15.0) * 6.0));
    p->moveto(-RADIUS/10, -RADIUS/20);
    p->lineto(RADIUS * 0.8, 0.0);
    p->lineto(-RADIUS/10, RADIUS/20);
    p->closepath();
    p->fill();
    p->restore();

    /* draw second hand */
    p->setrgbcolor(1.0, 0.0, 0.0);
    p->setlinewidth(2);
    p->save();
    p->rotate((float) -((ltime.tm_sec - 15.0) * 6.0));
    p->moveto(-RADIUS/5, 0.0);
    p->lineto(RADIUS, 0.0);
    p->stroke();
    p->restore();

    /* draw little circle at center */
    p->circle(0, 0, (float) RADIUS/30);
    p->fill();

    p->restore();

    p->end_page();

    p->close();

    exit(0);
    return (0);
}
